### ISSUE TITLE
Error "No module named 'py'" when running tests on a remote host using a socket server

### ISSUE THREAD

**Issue by acabarbaye** [pytest-dev/pytest-xdist#459]
In order to assess xdist when it comes to execution on remote host, I tried to do what is described here https://github.com/pytest-dev/pytest-xdist#sending-tests-to-remote-socket-servers with a really simple test as described below and received the following error in my client.
`
  File "<string>", line 1, in do_exec
  File "<remote exec>", line 13, in <module>
ModuleNotFoundError: No module named 'py'
`

- test:
```python
import unittest
from mypkg.a_file import print_stuff


class Test(unittest.TestCase):

    def test_it(self):
        print_stuff()
        self.assertTrue(True)
```

- project tree:
`
C:.
│   pytest.ini
│   setup.cfg
│
├───.cache
│   └───v
│       └───cache
│               lastfailed
│
├───.pytest_cache
│   │   .gitignore
│   │   CACHEDIR.TAG
│   │   README.md
│   │
│   └───v
│       └───cache
│               nodeids
│               stepwise
│
├───assets
│       style.css
│
├───mypkg
│   │   a_file.py
│   │   __init__.py
│   │
│   └───__pycache__
│           a_file.cpython-36.pyc
│           a_file.cpython-37.pyc
│           __init__.cpython-36.pyc
│           __init__.cpython-37.pyc
│
└───test
    │   __init__.py
    │
    ├───mypkg
    │   │   test_.py
    │   │   __init__.py
    │   │
    │   └───__pycache__
    │           test_.cpython-36-pytest-5.0.1.pyc
    │           test_.cpython-37-PYTEST.pyc
    │           __init__.cpython-36.pyc
    │           __init__.cpython-37.pyc
    │
    └───__pycache__
            __init__.cpython-36.pyc
            __init__.cpython-37.pyc
`
- pytest.ini
`
[pytest]
addopts = --html=remote-results.html --junitxml=remote-results.html --log-level=INFO --tx socket={myhost}:8888
`
- setup.cfg
`
[tool:pytest]
rsyncdirs = . mypkg 
`

- When running on my local machine:
`
λ python -m pytest 
===================================================== test session starts ====================================================== platform win32 -- Python 3.6.7, pytest-5.0.1, py-1.8.0, pluggy-0.12.0
rootdir: \test, inifile: pytest.ini
plugins: forked-1.0.2, html-1.21.1, metadata-1.8.0, xdist-1.29.0
collected 1 item

test\mypkg\test_.py .                                                                                                     [100%]

------------- generated xml file: remote-results.html -------------- --------- generated html file: file:\remote-results.html ---------- =================================================== 1 passed in 0.20 seconds ===================================================
`
- When running on a remote host with socketserver.py running there:
`
python -m pytest  -d --rsyncdir mypkg mypkg
.....

try: execmodel
except NameError:
   execmodel = get_execmodel('thread')
io = SocketIO(clientsock, execmodel)
io.write('1'.encode('ascii'))
serve(io, id='socket={myhost}:8888-slave')", line 1063, in executetask
  File "<string>", line 1, in do_exec
  File "<remote exec>", line 13, in <module>
ModuleNotFoundError: No module named 'py'

replacing crashed worker gw0

`

**Comment by RonnyPfannschmidt** [comment]
Execnet has no dependency management, so currently this only work on hosts that are prepared

**Comment by acabarbaye** [comment]
Could you be more specific?

> hosts that are prepared

How do I prepare the host?

**Comment by nicoddemus** [comment]
@acabarbaye unfortunately you have to do it manually: logging into the host via SSH, and installing the dependencies manually.

In modern setups (with several dependencies) unfortunately the remote execution features of pytest-xdist leave much to be desired.

**Comment by acabarbaye** [comment]
Could this be mentioned in the documentation?
Moreover, when installing on the remote host pytest and pytest-xdist, I get the following error:

> INTERNALERROR>   File "python36\current\lib\site-packages\pluggy\callers.py", line 187, in _multicall
> INTERNALERROR>     res = hook_impl.function(*args)
> INTERNALERROR>   File "python36\current\lib\site-packages\pytest_metadata\plugin.py", line 104, in pytest_testnodedown
> INTERNALERROR>     node.config._metadata.update(node.slaveoutput["metadata"])
> INTERNALERROR> KeyError: 'metadata'

**Comment by RonnyPfannschmidt** [comment]
thats a bug in pytest-metadata about remotes that dont have pytest-metadata installed

**Comment by acabarbaye** [comment]
I now installed `pytest-metadata` and the run seems to go through but no test is actually run:

> λ python -m pytest  -d --rsyncdir mypkg mypkg
> ====================================================================== test session starts ====================================================================== platform win32 -- Python 3.6.7, pytest-5.0.1, py-1.8.0, pluggy-0.12.0
> rootdir: C:\Users\adrcab01\OneDrive - Arm\Documents\GitHub\test\test, inifile: pytest.ini
> plugins: forked-1.0.2, html-1.21.1, metadata-1.8.0, xdist-1.29.0
> gw0 [0]
> 
> ------------------------------ generated xml file: \remote-results.xml ------------------------------- -------------------------- generated html file: file:\remote-results.html -------------------------- ================================================================= no tests ran in 0.70 seconds ==================================================================

**Comment by nicoddemus** [comment]
@RonnyPfannschmidt should we perhaps deprecate the remote test runs, so users don't waste too much time trying to diagnose problems on it?

As you have mentioned previously, I don't think it is feasible to "fix" it as it would need to introduce an entire new layer of dependency management.

**Comment by RonnyPfannschmidt** [comment]
In this case the rsync setting is missing as well

**Comment by RonnyPfannschmidt** [comment]
I wouldn't deprecate them, but it needs better setup instructions

**Comment by acabarbaye** [comment]
@RonnyPfannschmidt, I agree with you that the setup instructions could be a bit more explicit.
Regarding your comment, 
> In this case the rsync setting is missing as well
Could you tell me what I am actually missing which entails that no test is being picked up?

**Comment by RonnyPfannschmidt** [comment]
https://docs.pytest.org/en/3.0.1/xdist.html#sending-tests-to-remote-ssh-accounts <- `--rsyncdir`

**Comment by acabarbaye** [comment]
This is what I do
> python -m pytest -d --rsyncdir mypkg mypkg

**Comment by RonnyPfannschmidt** [comment]
and where are your tests?

**Comment by acabarbaye** [comment]
I see!
In my case, I have to put `--rsyncdir test test` instead of `--rsyncdir mypkg mypkg`. I misunderstood what the `rsyncdir` command was used for. 
From the following comment, I thought it was used to indicate the source code to rsync to the remote host but really, `mypkg` refers to the test package containing the tests
> Before running tests remotely, pytest efficiently “rsyncs” your program source code to the remote place.
I should have read this bit more carefully:
> Suppose you have a package mypkg which contains some tests that you can successfully run locally. 
Sorry for the bother.
It now works!

**Comment by RonnyPfannschmidt** [comment]
i see, this clearly indicates the dos could be much better however

thanks for the followus

**Comment by jlynchMicron** [comment]
@RonnyPfannschmidt @nicoddemus ,  although it would/could limit the features of pytest, do you think a possible to have a version of the remote unit test executor that only executes unit tests in the form of the python inbuilt unittest module?  That way, as long as the remote machine/subprocess has above python 2.1 installed, it could execute unittest tests controlled by pytest with no dependency needs.  This feature could be helpful for people that still keep most of their unit tests written in unittest style files.

**Comment by RonnyPfannschmidt** [comment]
its potentially possible (but really tricky, i would advise against trying) and not something i would want to work on in my own time
